[MRG+ ½] Fixing #7155 in stochastic_gradient.py

yl565 opened the pull request:
Fixes #7155 
Now calling predict_proba/predict_log_proba of SGDClassifier throw the correct error when not fitted and  calling predict_proba/predict_log_proba of GridSearchCV will not throw error when estimator=SGDClassifier and param_grid={'loss':['log']}

yl565 committed:
Fix #7155

raghavrv commented:
Thanks for the PR! Could you also add a non regression test please?

yl565 committed:
Fix test errors

yl565 commented:
Hi @raghavrv,  thanks for your comments! It is exciting that I'm submitting my first PR to sklearn.  I'm fairly new to the unit testing procedure. Do you mean add some test to the linear_model/tests/test_sgd.py and tests/test_grid_search.py for the bugs in #7155?

agramfort commented on sklearn/linear_model/stochastic_gradient.py:
```diff
@@ -714,9 +714,14 @@ def __init__(self, loss="hinge", penalty='l2', alpha=0.0001, l1_ratio=0.15,
+                            all_or_any=all)
+            is_fitted = True
+        except _NotFittedError:
+            is_fitted = False            
```
I am not sure it's the right fix.

is check_is_fitted really needed to know if predict_proba is supported?

can't you just remove the call to check_is_fitted ?

yl565 replied:
If I remove the call, GridSearchCV would  throw loss hinge error when predict_proba is called. The problem came from [this line](https://github.com/scikit-learn/scikit-learn/blob/master/sklearn/utils/metaestimators.py#L35). It tried to get the predict_proba attribute of SGDClassifier class but the default losss is hinge
In the first commit I pushed, I removed the call but it failed some tests:

raghavrv commented:
>  It is exciting that I'm submitting my first PR to sklearn.

Indeed it is ;)

> I'm fairly new to the unit testing procedure. Do you mean add some test to the `linear_model/tests/test_sgd.py` and `tests/test_grid_search.py` for the bugs in #7155?

Adding a test along the lines of

```py
def test_sgd_loss_param_grid_search():
    # Make sure the predict_proba works when loss is specified
    # as one of the parameters in the param_grid
    rng = np.random.RandomState(0)
    X = rng.random_sample([5, 2])
    y = rng.random_integers(0, 1, [5, ])
    param_grid = {
        'loss': ['log'],
    }
    clf_grid = grid_search.GridSearchCV(estimator=SGDClassifier(random_state=42),
                                        param_grid=param_grid)
    clf_grid.fit(X, y).predict_proba(X)
```

added to `linear_model/tests/test_sgd.py` would be good. (This is an issue with sgd (as we are hacking it a bit) and hence should go into sgd tests.)

yl565 committed:
add_tests

yl565 commented:
@raghavrv Your instruction is really helpful! I added the test along with a test to make sure when predict_proba is called the error raised is `not fitted` but not `AttributeError: probability estimates are not available for loss='hinge'`. Now running `test_sgd.py` with the original `stochastic_gradient.py` would give the following errors:
```python
======================================================================
ERROR: test_loss_param_grid_search (sklearn.linear_model.tests.test_sgd.DenseSGDClassifierTestCase)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/data/Dropbox/Python/scikit-learn/sklearn/linear_model/tests/test_sgd.py", line 836, in test_loss_param_grid_search
    clf_grid.predict_proba(X)
  File "/home/data/Dropbox/Python/scikit-learn/sklearn/utils/metaestimators.py", line 35, in __get__
    self.get_attribute(obj)
  File "/home/data/Dropbox/Python/scikit-learn/sklearn/linear_model/stochastic_gradient.py", line 756, in predict_proba
    self._check_proba()
  File "/home/data/Dropbox/Python/scikit-learn/sklearn/linear_model/stochastic_gradient.py", line 721, in _check_proba
    " loss=%r" % self.loss)
AttributeError: probability estimates are not available for loss='hinge'

======================================================================
ERROR: test_not_fitted_error (sklearn.linear_model.tests.test_sgd.DenseSGDClassifierTestCase)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/data/Dropbox/Python/scikit-learn/sklearn/linear_model/tests/test_sgd.py", line 846, in test_not_fitted_error
    clf.predict_proba, X)
  File "/home/data/Dropbox/Python/scikit-learn/sklearn/linear_model/stochastic_gradient.py", line 756, in predict_proba
    self._check_proba()
  File "/home/data/Dropbox/Python/scikit-learn/sklearn/linear_model/stochastic_gradient.py", line 721, in _check_proba
    " loss=%r" % self.loss)
AttributeError: probability estimates are not available for loss='hinge'

======================================================================
ERROR: test_loss_param_grid_search (sklearn.linear_model.tests.test_sgd.SparseSGDClassifierTestCase)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/data/Dropbox/Python/scikit-learn/sklearn/linear_model/tests/test_sgd.py", line 837, in test_loss_param_grid_search
    clf_grid.predict_log_proba(X)
  File "/home/data/Dropbox/Python/scikit-learn/sklearn/utils/metaestimators.py", line 35, in __get__
    self.get_attribute(obj)
  File "/home/data/Dropbox/Python/scikit-learn/sklearn/linear_model/stochastic_gradient.py", line 822, in predict_log_proba
    self._check_proba()
  File "/home/data/Dropbox/Python/scikit-learn/sklearn/linear_model/stochastic_gradient.py", line 721, in _check_proba
    " loss=%r" % self.loss)
AttributeError: probability estimates are not available for loss='hinge'

======================================================================
ERROR: test_not_fitted_error (sklearn.linear_model.tests.test_sgd.SparseSGDClassifierTestCase)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/data/Dropbox/Python/scikit-learn/sklearn/linear_model/tests/test_sgd.py", line 846, in test_not_fitted_error
    clf.predict_proba, X)
  File "/home/yichuanliu/anaconda3/envs/scikit-learn-dev/lib/python3.5/unittest/case.py", line 1258, in assertRaisesRegex
    return context.handle('assertRaisesRegex', args, kwargs)
  File "/home/yichuanliu/anaconda3/envs/scikit-learn-dev/lib/python3.5/unittest/case.py", line 176, in handle
    callable_obj(*args, **kwargs)
  File "/home/data/Dropbox/Python/scikit-learn/sklearn/linear_model/tests/test_sgd.py", line 44, in predict_proba
    return super(SparseSGDClassifier, self).predict_proba(X)
  File "/home/data/Dropbox/Python/scikit-learn/sklearn/linear_model/stochastic_gradient.py", line 756, in predict_proba
    self._check_proba()
  File "/home/data/Dropbox/Python/scikit-learn/sklearn/linear_model/stochastic_gradient.py", line 721, in _check_proba
    " loss=%r" % self.loss)
AttributeError: probability estimates are not available for loss='hinge'

----------------------------------------------------------------------
Ran 168 tests in 1.292s

FAILED (errors=4)
```

nelson-liu commented on sklearn/linear_model/stochastic_gradient.py:
```diff
@@ -32,7 +32,7 @@
 from .sgd_fast import EpsilonInsensitive
 from .sgd_fast import SquaredEpsilonInsensitive
-
+from ..exceptions import NotFittedError as _NotFittedError
```
i think you should add a newline after this, to replace the one you removed.

nelson-liu commented on sklearn/linear_model/tests/test_sgd.py:
```diff
@@ -20,8 +20,9 @@
 from sklearn.base import clone
 from sklearn.linear_model import SGDClassifier, SGDRegressor
 from sklearn.preprocessing import LabelEncoder, scale, MinMaxScaler
-
```
same here, insert a newline to replace the one you removed

nelson-liu commented on sklearn/linear_model/tests/test_sgd.py:
```diff
@@ -823,7 +824,27 @@ def test_multiple_fit(self):
         y = [["ham", "spam"][i] for i in LabelEncoder().fit_transform(Y)]
         clf.fit(X[:, :-1], y)
 
-
```
here as well

nelson-liu commented on sklearn/linear_model/tests/test_sgd.py:
```diff
@@ -823,7 +824,27 @@ def test_multiple_fit(self):
+        # Make sure raise not fitted error and not the loss='hinge' error
+        clf = self.factory()
+        assert_raises_regexp(_NotFittedError,
+                             "This {} instance is not fitted yet. "
```
python 2.6 requires numbered fields with `format`, hence why travis is a bit angry

yl565 committed:
fix_py26_error

yl565 committed:
add spaces

raghavrv commented on sklearn/linear_model/stochastic_gradient.py:
```diff
@@ -757,6 +764,7 @@ def predict_proba(self):
         return self._predict_proba
 
     def _predict_proba(self, X):
+        check_is_fitted(self, ["t_", "coef_", "intercept_"], all_or_any=all)
```
Now this gets checked  twice no? Do we want that?

yl565 commented:
@nelson-liu Thanks for the comments! Newlines added.

raghavrv commented:
Also why do we even set the fit params like `t_` at init? I think we should fix that too.

Additionally I think we could simplify this by combining the `_check_proba`, `_predict_proba` and `predict_proba` into a single method with a decorator. The decorator would be similar to the `if_delegete_has_method` decorator. It can be made to check the loss and raise `AttributeError`. That was the fix I had in mind. But I am not sure if that would work as I haven't tried it yet... Maybe you can try that. That will make the code more readable I feel...

yl565 committed:
use decorator, do not set fit params at init

yl565 committed:
fix coef_ initialization error raised in passive_aggressive.py

yl565 commented:
@raghavrv Following your idea, I used decorator on a single method `predict_proba`. Also removed setting `t_` and `coef_` at init. Seems to be working as expected now?

raghavrv commented on sklearn/linear_model/stochastic_gradient.py:
```diff
@@ -496,7 +501,7 @@ def partial_fit(self, X, y, classes=None, sample_weight=None):
         -------
         self : returns an instance of self.
-        """
+        """            
```
trailing white space

yl565 committed:
fix pep8 and redundent _CheckProba in predict_log_proba

yl565 commented:
@raghavrv PEP8 problem fixed!

raghavrv commented:
Thanks! When `loss='hinge'` we would want an `AttributeError` instead of a `NotFittedError` error I think...

Before further reviews - @MechCoder and @agramfort does this approach seem okay?

raghavrv commented:
@yl565 Could you revert back the PEP8 changed unrelated to this PR please?

yl565 commented:
@raghavrv Thanks for all your help and comments! Will revert the PEP8 unrelated to this PR later tonight.
> When loss='hinge' we would want an AttributeError instead of a NotFittedError error I think when the estimator is not fitted yet...

I have tried that but it will give the following testing error: 
```python
======================================================================
FAIL: sklearn.tests.test_common.test_non_meta_estimators('SGDClassifier', <class 'sklearn.linear_model.stochastic_gradient.SGDClassifier'>)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/yichuanliu/anaconda3/envs/scikit-learn-dev/lib/python3.5/site-packages/nose/case.py", line 198, in runTest
    self.test(*self.arg)
  File "/home/data/Dropbox/Python/scikit-learn/sklearn/utils/testing.py", line 343, in wrapper
    return fn(*args, **kwargs)
  File "/home/data/Dropbox/Python/scikit-learn/sklearn/utils/estimator_checks.py", line 1060, in check_estimators_unfitted
    est.predict_proba, X)
  File "/home/data/Dropbox/Python/scikit-learn/sklearn/utils/testing.py", line 424, in assert_raise_message
    (message, error_message))
AssertionError: Error message does not include the expected string: 'fit'. Observed error message: "probability estimates are notavailable for loss='hinge'
```

raghavrv commented:
While [initializing](https://github.com/scikit-learn/scikit-learn/blob/master/sklearn/utils/estimator_checks.py#L255) the `SGDClassifier` for that test, you will have to special case and set `loss='log'`...

BTW can you check why when `loss='hinge'`, [this](https://github.com/scikit-learn/scikit-learn/blob/master/sklearn/utils/estimator_checks.py#L1058) line passes and invokes the test?

raghavrv commented:
I was thinking along the lines of 

```py

from sklearn.model_selection import GridSearchCV
from sklearn.exceptions import NotFittedError
from sklearn.utils.validation import check_is_fitted

class _CheckProba(object):
    def __init__(self, fn):
        self.fn = fn

    def __call__(self, *args, **kw):
        return self.fn(*args, **kw)

    def __get__(self, obj, type=None):
        if obj.loss not in ("log", "modified_huber"):
            raise AttributeError("probability estimates are not available for"
                                 " loss=%r" % obj.loss)
        return self.__class__(self.fn.__get__(obj, type))

class Estim(object):
    def __init__(self, loss='hinge'):
        self.loss=loss
        
    @_CheckProba
    def predict_proba(self, X):
        check_is_fitted(self, 'coef_')
        return np.arange(X.shape[0])
    
    def fit(self, X, y=None):
        self.coef_ = 1
        return self
    
    def __repr__(self):
        return "Estim(loss=%r)" % self.loss
```

Test the following ... 

```py    
>>> X = [[1, 2], [2, 3]]
>>> y = [0, 1]
>>> hasattr(Estim().fit(X, y), 'predict_proba')
False
>>> hasattr(Estim(loss='log').fit(X, y), 'predict_proba')
True
>>> Estim(loss='log').predict_proba(X)
NotFittedError
```

yl565 commented:
@raghavrv 
```python
    def __get__(self, obj, type=None):
        if obj.loss not in ("log", "modified_huber"):
            raise AttributeError("probability estimates are not available for"
                                 " loss=%r" % obj.loss)
```
The `loss=hinge` exception would still raise if not checking is_fitted first. When calling predict_proba of GridSearchCV [`@if_delegate_has_method(delegate='estimator')`](https://github.com/scikit-learn/scikit-learn/blob/master/sklearn/grid_search.py#L444) will try [get attribute](https://github.com/scikit-learn/scikit-learn/blob/master/sklearn/utils/metaestimators.py#L26) of `estimator.predict_proba` but the estimator has default `loss=hinge`.

yl565 commented:
@raghavrv 
>While initializing the SGDClassifier for that test, you will have to special case and set loss='log'

It seems that the [testing error](https://github.com/scikit-learn/scikit-learn/pull/7159#issuecomment-238591260) is still there after I set `loss='log'` in [`set_testing_parameters`](https://github.com/scikit-learn/scikit-learn/blob/master/sklearn/utils/estimator_checks.py#L255). Apparently [`check_estimators_unfitted`](https://github.com/scikit-learn/scikit-learn/blob/master/sklearn/utils/estimator_checks.py#L1035) is not calling `set_testing_parameters`. If I add a line `set_testing_parameters(est)` in `check_estimators_unfitted` the SGDClassifier will pass the test but I'm not sure if this is going to affect other estimators. Any suggestions?

raghavrv commented:
Ah okay I got the issue! This is similar to #4902... I think @jnothman will again know the best thing to do here...

I have a feeling that we should do away with the hack in `SGDClassifier` completely. Maybe just raise a simple `ValueError` when `predict_proba` is called for `loss='hinge'`, and let the user know that `loss='hinge'` is not supported instead of trying to make `hasattr`/`getattr` work. But please do wait for @jnothman's comment before proceeding!

agramfort commented:
ping @jnothman

dhimmel commented:
I'm also interested in seeing a solution implemented. Double ping @jnothman :smiley_cat:

jnothman commented:
Sorry I didn't see the ping. I've now set up some GMail labels to tell me when I'm mentioned, so in the future you have a hope!

Firstly: what a mess. Not to your PR, but the bug!

But your PR, unfortunately, takes the wrong tack. The fault here is with `GridSearchCV` not with `SGD*`. The `hasattr` delegation in `GridSearchCV` assumed that a method was available in the un-fitted base estimator iff it would be available in the fitted estimator. This is wrong as long as we're using the same sort of magic to make `predict_proba` (dis)appear as a function of parameters.

I am not sure we can find a solution that always works when `GridSearchCV` has not yet been fit. But it is clear now that `hasattr` in `GridSearchCV` should be delegated to `best_estimator_` if that exists, and to `estimator` only if it does not. So I suppose we shall have to change `if_delegate_has_method` to incorporate a back-off strategy.

What a mess!

yl565 committed:
Revert to master

yl565 committed:
Implement solution suggested by @jnothman

I changed `if_delegate_has_method` to accept a tuple of `delegate`. AttributeError will only raise if the method is not in any of the delegates.
I then decorated `predict_proba`, `predict`, etc. in `BaseSearchCV` with `@if_delegate_has_method(delegate=('best_estimator_', 'estimator'))`
Tests have been added

yl565 committed:
Revert estimator_checks.py to master

raghavrv commented:
> So I suppose we shall have to change if_delegate_has_method to incorporate a back-off strategy.

+1. I think I suggested something similar in the other PR [here](https://github.com/scikit-learn/scikit-learn/issues/4902#issuecomment-117420538)...

raghavrv commented on sklearn/tests/test_grid_search.py:
```diff
@@ -673,6 +674,23 @@ def test_grid_search_with_multioutput_data():
+    X = np.arange(20).reshape(5, -1)
+    y = [0, 0, 1, 1, 1]
+    clf= GridSearchCV(estimator=SGDClassifier(),
+                            param_grid=param_grid)
```
pep8

raghavrv commented on sklearn/tests/test_grid_search.py:
```diff
@@ -673,6 +674,23 @@ def test_grid_search_with_multioutput_data():
+    clf= GridSearchCV(estimator=SGDClassifier(),
+                            param_grid=param_grid)
+
+    # when not fitted, predict_proba diabled because default loss='hinge'
```
disabled

raghavrv replied:
I'd write it as

```
When the estimator is not fitted, `predict_proba` is not available as the default loss is 'hinge'.
```

yl565 commented:
@jnothman and @raghavrv, in my latest commits, I modified  `If_delegate_has_method` to accept a tuple of delegates. Exception will only raise if all of them do not contain the method. Methods in `BaseSearchCV` has been changed to:
```python
@If_delegate_has_method(delegate = ('best_estimator_',  'estimator'))
def predict_proba(self, X):
```
An alternative but harder to implement solution I can think of is to pass a logical expression, e.g.:
```python
@If_delegate_has_method(delegate = 'best_estimator_ or estimator')
def predict_proba(self, X):
```

raghavrv commented on sklearn/utils/metaestimators.py:
```diff
@@ -10,11 +10,47 @@
+
+    >>> class LayerC(object):
+    ...     def __init__(self):
+    ...         self.c = '123'
```
Sorry I fail to understand. When does this happen and how is it related to this PR?

raghavrv commented on sklearn/utils/metaestimators.py:
```diff
@@ -10,11 +10,47 @@
 
 
+def hasattr_nested(obj, attr):
+    """Check if obj has a (possibly nested) attribute
```
Sorry I fail to understand. When does this happen and how is it related to this PR?

yl565 replied:
`hasattr` doesn't work with nested attribute. `hasattr_nested` is like `attrgetter`, but only check if attribute is available and not accessing the attribute
In this case we are trying to check if the following statement is true:
```Python
hasattr_nested(GridSearchCV(), 'estimator.predict_proba')
```

raghavrv replied:
Okay. The docstring is too verbose. This function will probably not be used anywhere else. Can you just describe what it does in the docstring without examples?

raghavrv commented:
> An alternative but harder to implement solution I can think of is to pass a logical expression, e.g.:

No the current approach seems okay to me.

yl565 committed:
Fix pep8 and modify comments as @raghavrv suggested

yl565 commented:
@raghavrv pep8 and problems in comments fixed!

raghavrv commented on sklearn/utils/metaestimators.py:
```diff
@@ -10,11 +10,51 @@
 __all__ = ['if_delegate_has_method']
 
 
+def hasattr_nested(obj, attr):
```
--> `_hasattr_nested`

raghavrv commented:
That's it for me...

yl565 committed:
Simplified docstring of `hasattr_nested`and renamed it `_hasattr_nested`

jnothman commented:
1. That's not what I suggested. If `best_estimator_` exists, then the presence or absence of an attribute on `estimator` is irrelevant (unless the grid search is about to be `fit` again and has changed parameters, but for now we can assume that case is pathological).
2. You will need to add tests in `sklearn/utils/tests/test_metaestimators.py`, I think

jnothman commented:
do you get what I'm suggesting, @yl565? I think it's an important difference.

yl565 committed:
Moving tests and ignore other delegates if the attribute has already been found

yl565 commented on sklearn/utils/metaestimators.py:
```diff
@@ -32,7 +52,11 @@ def __get__(self, obj, type=None):
+                if _hasattr_nested(obj, item):
+                    break
+            else:
+                attrgetter(self.attribute_name[-1])(obj)
```
@jnothman I modified the code to break and ignore other (later) delegates if the attribute has already been found. Does this implemented your suggestion?

jnothman replied:
No, because you're looking up `best_estimator_.predict`, for instance, and if `best_estimator_` exists but its `predict` does not, you're backing off to `estimator`, when you should be raising an `AttributeError`.

yl565 committed:
Changed from `tuple([delegate])` to `(delegate,)`

amueller commented on sklearn/utils/metaestimators.py:
```diff
@@ -10,11 +10,30 @@
 __all__ = ['if_delegate_has_method']
 
 
+def _hasattr_nested(obj, attr):
```
hm I would have thought operator.attrgetter  would solve this, but I guess it doesn't. That's kinda weird.

amueller replied:
I mean you could do a try except.

yl565 replied:
@amueller I haven't tried implement this using `try except attrgetter` because I'm worried it may affect performance. Do you think it would be better than the current approach?
```Python
    def __get__(self, obj, type=None):
        # raise an AttributeError if the attribute is not present on the object
        if obj is not None:
            # delegate only on instances, not the classes.
            # this is to allow access to the docstrings.
            ecpt = None
            for item in self.attrs_tuple:
                try:
                    attrgetter(item)(obj)
                    break
                except ValueError as e:
                    ecpt = e
            else:
                raise(ecpt)
```

amueller commented on sklearn/utils/metaestimators.py:
```diff
@@ -10,11 +10,30 @@
+    if the all items in ``attribute_name`` is not present on the base object.
+    attribute_name can be a single string or a tuple of strings. The
+    ``AttributeError`` raised will indicate the last item is not present on
+    the base object
```
full stop.

amueller commented on sklearn/utils/metaestimators.py:
```diff
@@ -23,7 +42,8 @@ class _IffHasAttrDescriptor(object):
     def __init__(self, fn, attribute_name):
         self.fn = fn
-        self.get_attribute = attrgetter(attribute_name)
+        self.attribute_name = attribute_name
```
This is now a tuple. The variable name should reflect this.

jnothman replied:
I.e. should be `attr_names` or similar

jnothman commented on sklearn/utils/metaestimators.py:
```diff
@@ -43,30 +67,60 @@ def __get__(self, obj, type=None):
 def if_delegate_has_method(delegate):
     """Create a decorator for methods that are delegated to a sub-estimator
 
+    Delegate can be string or a tuple of strings
```
Semantics should be detailed.

yl565 committed:
Check first if the first item in delegate is an attribute of object
If yes get method as attribute of the delegate, if no check the next item in delegate

jnothman commented:
Noting that this bug is a regression I think we should consider it a blocker for 0.18, @amueller.

jnothman commented on sklearn/utils/metaestimators.py:
```diff
@@ -32,7 +37,13 @@ def __get__(self, obj, type=None):
             # delegate only on instances, not the classes.
             # this is to allow access to the docstrings.
-            self.get_attribute(obj)
+            for item in self.delegate_names:
```
"delegate" rather than "item"

jnothman replied:
or `delegate_name`

jnothman commented on sklearn/utils/metaestimators.py:
```diff
@@ -43,30 +54,61 @@ def __get__(self, obj, type=None):
+    True
     """
-    return lambda fn: _IffHasAttrDescriptor(fn, '%s.%s' % (delegate, fn.__name__))
+    if not isinstance(delegate, tuple):
```
I think list should also be supported.

jnothman commented on sklearn/utils/metaestimators.py:
```diff
@@ -32,7 +37,13 @@ def __get__(self, obj, type=None):
             # this is to allow access to the docstrings.
-            self.get_attribute(obj)
+            for item in self.delegate_names:
+                if hasattr(obj, item):
```
If the point of using attrgetter before was to allow the delegate to have '.' in it, then we'd better maintain that, with something like:
```python
try:
    delegate = attrgetter(delegate_name)(obj)
except AttributeError:
    continue
else:
    getattr(delegate, self.method_name)
    break
```

yl565 committed:
Add support for list. Use attrgetter instead of hasattr

yl565 commented:
@jnothman The suggested changes has been made! I also simplified the docstring.

amueller commented:
When did this work @jnothman?

jnothman commented:
When we didn't do hasattr delegation for `GridSearchCV`. I've not checked when.

yl565 committed:
Modify docstring, simplify docstring example

jnothman commented:
I think it's been broken since 0.16, @amueller (due to #3982)

jnothman commented on sklearn/utils/metaestimators.py:
```diff
@@ -80,22 +80,22 @@ def if_delegate_has_method(delegate):
+    ...     def predict(self):
+    ...         pass
     ...
+    >>> class MetaEstTestTuple(MetaEst):
```
I would rather this kind of variation and testing to be in unit tests in a different file, not doctests.

lesteve replied:
+1

jnothman commented on sklearn/utils/tests/test_metaestimators.py:
```diff
@@ -24,3 +30,21 @@ def test_delegated_docstring():
+    clf = GridSearchCV(estimator=SGDClassifier(),
+                       param_grid=param_grid)
+
+    # When the estimator is not fitted, `predict_proba` is not available as the default loss is 'hinge'.
```
PEP8 line length (required <80 chars)

jnothman commented on sklearn/utils/tests/test_metaestimators.py:
```diff
@@ -24,3 +30,21 @@ def test_delegated_docstring():
+    assert_false(hasattr(clf, "predict_proba"))
+    clf.fit(X, y)
+    clf.predict_proba(X)
+    clf.predict_log_proba(X)
```
Just to be sure, please assert that when `param_grid={'loss':['hinge']}`, `clf.predict_proba` is not available.

jnothman commented on sklearn/utils/tests/test_metaestimators.py:
```diff
@@ -24,3 +30,21 @@ def test_delegated_docstring():
+    }
+    X = np.arange(20).reshape(5, -1)
+    y = [0, 0, 1, 1, 1]
+    clf = GridSearchCV(estimator=SGDClassifier(),
```
explicitly set `loss='hinge'` here so the test is a little clearer.

lesteve commented on sklearn/utils/metaestimators.py:
```diff
@@ -14,16 +14,22 @@ class _IffHasAttrDescriptor(object):
     descriptors.
     """
-    def __init__(self, fn, attribute_name):
+    def __init__(self, fn, delegate_names, method_name):
```
I would keep the old name `attribute_name` rather than `method_name` since this class can work with any kind of attribute. Also update the docstring accordingly.

yl565 committed:
Move doctests to unit tests, add more tests, rename `method_name`, fix pep8

yl565 committed:
remove unused import

jnothman commented:
Thanks @yl565 for all the work, despite reviews being slow. I think your code is exactly what we've sought.

However, TBH, I'm not altogether happy with this solution: the extension to `if_delegate_has_method` makes it inelegant, and `GridSearchCV` still may give the wrong response if we're trying to do type sniffing before fitting. Then this makes problems elsewhere: if a `GridSearchCV` instance is the estimator underlying some other metaestimator, say `RFECV` then it similarly needs to delegate to its fitted estimator `estimator_` instead of the input parameter `estimator`.

Ultimately, all this suggests that (for some methods at least) duck type sniffing should only ever be performed after fitting (and should be delegated to `best_estimator_` and similar only). In practice, validation is performed early to avoid expensive fitting.

I'd be okay with merging this, given another core dev's approval, but we should look into making things more robust...?

yl565 committed:
Alternative solution with check_is_fitted

yl565 committed:
Merge branch 'fix_7155_alt_check_fitted' into fix_7155

jnothman commented:
CI is failing because you introduced PEP8 errors. Requiring estimators to be fitted before we do type sniffing is actually a bigger problem than here. We have to go looking for wherever we (or indeed, others) use hasattr or get attributes to determine capabilities before fitting. It's no small task, I think.

yl565 committed:
Use new function if_fitted_delegate_has_method

yl565 commented:
@jnothman Thanks for the review. It's fun solving the problems and I enjoy the learning process. If we could make things more robust I would definitely want to give it a shot.  I'm testing the alternative solution following your suggestion. To (temporarily) improve compatibility, I made `if_delegate_has_method` behave the same as before and used a new function `if_fitted_delegate_has_method` that checks if the meta-estimator has been fitted before duck typing. I think maybe we can start here and then go looking for where hasattr are used to see if this solution is robust?

yl565 committed:
Fix pep8

lesteve commented on sklearn/utils/metaestimators.py:
```diff
@@ -32,41 +37,52 @@ def __get__(self, obj, type=None):
+        fn_check_obj(obj) with obj as the metaestimator instance.
+        The function will be executed before ducktyping. Typically used for
+        checking if the estimator has been fitted.
+
```
I would leave the doctests as they were or add them as tests to sklearn/utils/tests/test_metaestimators.py.

You could also add additional tests in the same file as you did at some point of this PR.

lesteve commented on sklearn/utils/tests/test_metaestimators.py:
```diff
@@ -24,3 +33,33 @@ def test_delegated_docstring():
                 in str(MockMetaEstimator().func.__doc__))
+
+
+def test_stochastic_gradient_loss_param():
```
This test doesn't really belong to test_meta_estimators.py maybe somewhere like sklearn/tests/test_grid_search.py would be better.

ogrisel replied:
+1 for moving it to `sklearn/model_selection/tests/test_search.py`

amueller commented:
Only being able to sniff out supported methods after fitting seems to be somewhat of a problem :( This "only" happens with GridSearchCV and RandomizedSearchCV, right?
This seems to have weird impact on the estimator tags (#6599), because they are now only defined after fitting... Basically all possible tags can change during fit based on what steps a grid-search put into a pipeline.

jnothman commented:
It happens when capabilities can't be determined from parameters alone. Currently that means `GridSearchCV` and `RandomizedSearchCV`, I think.

jnothman commented:
Other estimators which could delegate to a `GridSearchCV` may be a problem too. A pipeline with a gridsearch as its last step, for instance, has indeterminate `predict_proba` support until after fitting.

Then in terms of duck-typing I see two choices:
* only ever test for attributes after fitting
* specify a set of attributes that can be tested for before fitting, the remainder only after. The same may be said of tags.

For reference, here is a quick-and-dirty listing of attributes we hasattr (on things that mostly aren't data):

```
1	"Lock"
1	"_idf_diag"
1	"_optimizer"
1	"alpha"
1	"alphas"
1	"args"
1	"compute_labels"
1	"connected_components"
1	"f_cached"
1	"fit_predict"
1	"iteritems"
1	"kind"
1	"lsqr"
1	"min_weight_fraction_leaf"
1	"n_nonzero_coefs"
1	"penalty"
1	"predict_log_proba"
1	"random_state"
1	"score"
1	"sorted_indices"
1	"sparsify"
1	"split"
1	"toarray"
1	attr
1	hasattr(MetaEst(HasNoPredict(
1	hasattr(MetaEst(HasPredict(
1	reg
2	"dtype"
2	"get_feature_names"
2	"get_params"
2	"item"
2	"items"
2	"kernel"
3	"data"
3	"fit_transform"
3	"l1_ratio"
3	"n_iter"
3	"transform"
3	method
4	"fit"
4	"partial_fit"
5	"decision_function"
5	"n_clusters"
5	"predict"
6	"n_components"
8	"predict_proba"
```

For example: `predict_proba` is a "only after fit" case. I think `predict` *should* be stable over fitting.

`grep -oR 'hasattr([^\)]*)' sklearn/ | grep -v tests | grep -v external | grep -v '_\b' | cut -d: -f2 | grep -v 'hasattr(\(X\|t\|x\|v\|y\|np\|self.init\|f\|filename\|alg\|M\|yval\|w\|self.cv\)\b' | grep -v '^Binary' | cut -d' ' -f2 | tr "'" "\"" | sed 's/)$//' | count | sort -n`

yl565 commented:
So maybe we apply `@if_delegate_has_method` to stable method like `predict` and apply `@if_fitted_delegate_has_method` to `predict_proba`? Or we specify in `metaestimators.py` a list of attributes that should be checked if fitted?

amueller commented:
Sooo we want to warn people if they check ``hasattr`` before fit on the ones that are not stable?

``predict`` can be not stable if I grid-search a Pipeline that can have as steps either ``PCA`` or SVC``.

jnothman commented:
More practically, can we separate the long-term solution from one which
supports the case broken in this bug, but does not break anything new? I
think the answer is the back-off delegation that was implemented in this PR
a few days ago...

On 8 September 2016 at 01:58, Yichuan Liu <notifications@github.com> wrote:

> So maybe we apply @if_delegate_has_method to stable method like predict
> and apply @if_fitted_delegate_has_method to predict_proba? Or we specify
> in metaestimators.py a list of attributes that should be checked if
> fitted?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/7159#issuecomment-245328684>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAEz66jUOpqd4w5lYdzAOJNqYvoxPmwOks5qnt8KgaJpZM4Jek2m>
> .
>

yl565 commented:
@jnothman Do you think it is better to revert back to 265fc70?

amueller commented:
@yl565 I think that was the suggestion.

yl565 committed:
revert back to 265fc70

amueller commented on sklearn/utils/metaestimators.py:
```diff
@@ -14,16 +14,22 @@ class _IffHasAttrDescriptor(object):
-    This allows ducktyping of the decorated method based on ``attribute_name``.
+    This allows ducktyping of the decorated method based on
+    ``delegate.attribute_name`` where ``delegate`` is the first item in
+    ``delegate_names`` that is an attribute of the base object.
```
I find this sentence hard to understand. maybe make it two sentences. And say "Here  ``delegate`` is the first item in ``delegate_names`` for which ``hasattr(object, delegate)`` is true? not sure.

yl565 replied:
How about the following:

Using this class to create a decorator for a class method ``object.attribute_name``.
An ``AttributeError`` will be raised if none of the delegates (specified in
``delegate_names``) is an attribute of ``object`` (i.e. ``hasattr(object, delegate )``
is ``False``) or none of the delegates has an attribute ``attribute_name`` (i.e.
``hasattr(delegate, attribute_name)`` is ``False``).

This allows ducktyping of the decorated method ``object.attribute_name`` based on
``object.delegate.attribute_name`` where ``delegate`` is the first item in
``delegate_names`` for which ``hasattr(object, delegate )`` is ``True``).

ogrisel replied:
I don't like " for a class method `object.attribute_name`":

- `attribute_name` can be any attribute, not necessarily a method
- a class method named "object.attribute_name" is weird or even contradictory.

ogrisel replied:
Let's just spit the second sentence as @amueller suggested.

ogrisel commented on sklearn/utils/metaestimators.py:
```diff
@@ -32,7 +38,17 @@ def __get__(self, obj, type=None):
+                    continue
+                else:
+                    getattr(delegate, self.attribute_name)
+                    break
```
Could you please add a inline comment to explain in which case(s) using getattr as fallback is required?

ogrisel commented:
Besides the unaddressed comments still visible in the diff view of this PR on github, LGTM.

jnothman commented on sklearn/utils/metaestimators.py:
```diff
@@ -14,16 +14,22 @@ class _IffHasAttrDescriptor(object):
     Using this class to create a decorator will raise an ``AttributeError``
-    if the ``attribute_name`` is not present on the base object.
+    if none of the delegates (specified in ``delegate_names``) is an attribute
+    of the base object or none of the delegates has an attribute
```
"none of the delegates has" -> "the first found delegate does not have"

ogrisel commented:
@yl565 let me finish this PR for you. We want to have it merged into based before cutting the 0.18.X release branch tonight.

yl565 commented:
@ogrisel Sorry for couldn't response earlier, have been busy preparing a paper. Please go head. Thanks.

ogrisel commented:
No problem!

ogrisel commented:
#7401 has been merged. Thanks for the fix @yl565!
